transport-native-epoll/src/main/c/io_netty_channel_libaio_DirectFileDescriptorController.c

```diff
@@ -0,0 +1,347 @@
+    }
+
+    io_prep_pwrite(iocb, fileHandle, getBuffer(env, bufferWrite), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT void JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_submitRead
+  (JNIEnv * env, jclass clazz, jint fileHandle, jobject contextPointer, jlong position, jint size, jobject bufferRead, jobject callback) {
+
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    struct iocb * iocb = getIOCB(theControl);
+
+    if (iocb == NULL) {
+       throwIOException(env, "Not enough space on the queue for submitting");
+       return;
+    }
+
+    io_prep_pread(iocb, fileHandle, getBuffer(env, bufferRead), size, position);
+    iocb->data = (void *) (*env)->NewGlobalRef(env, callback);
+
+
+    submit(env, theControl->ioContext, iocb);
+}
+
+JNIEXPORT jint JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_poll
+  (JNIEnv * env, jobject obj, jobject contextPointer, jobjectArray callbacks, jint min, jint max) {
+
+    int i = 0;
+    struct io_control * theControl = getIOControl(env, contextPointer);
+
+    int result = io_getevents(theControl->ioContext, min, max, theControl->events, 0);
+
+    for (i = 0; i < result; i++) {
+        struct iocb * iocbp = (struct iocb *)theControl->events[i].obj;
+
+        int eventResult = theControl->events[i].res;
+
+        #ifdef DEBUG
+            fprintf (stderr, "Poll res: %d totalRes=%d\n", eventResult, result);
+        #endif
+
+        if (eventResult < 0) {
+            #ifdef DEBUG
+                fprintf (stderr, "Error: %s\n", strerror(-eventResult));
+            #endif
+
+            jstring jstrError = (*env)->NewStringUTF(env, strerror(-eventResult));
+
+            jobject errorObject = (*env)->NewObject(env,
+                                                   theControl->errorInfoClass,
+                                                   theControl->errorInfoConstr,
+                                                   (jobject)(iocbp->data),
+                                                   (jint)(-eventResult),
+                                                   jstrError);
+
+            if (errorObject == NULL) {
+                return -1;
+            }
+
+            (*env)->SetObjectArrayElement(env, callbacks, i, errorObject);
+        }
+        else {
+            (*env)->SetObjectArrayElement(env, callbacks, i, (jobject)iocbp->data);
+        }
+
+        if (iocbp->data != NULL) {
+            (*env)->DeleteGlobalRef(env, (jobject)iocbp->data);
+        }
+
+        putIOCB(theControl, iocbp);
+    }
+}
+
+JNIEXPORT jobject JNICALL Java_io_netty_channel_libaio_DirectFileDescriptorController_newAlignedBuffer
+(JNIEnv * env, jclass clazz, jint size, jint alignment) {
+    if (size % alignment) {
+        throwRuntimeException(env, "Buffer size needs to be aligned to passed argument");
+        return 0;
```

Comment:
It doesn't really matter, these are after a throwRuntimeException call, the java method will signal an exception and these will just be ignored.

I will set it to NULL just for consistency.